events/events.go

```diff
@@ -31,9 +34,100 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
+}
+
 // EventRegistrar is an object that you can register EventListeners on.
 type EventRegistrar interface {
-	RegisterListener(EventListener)
+	AddListener(EventListener)
+	RemoveListener(EventListener)
+}
+
+type baseEventRegistrar struct {
+	lock      sync.RWMutex
+	listeners []EventListener
```

Comment:
I think we have to add a comment saying how listeners is used when adding, removing and iterating. Also that it should never be mutated (only reassigned). I feels a bit fragile in the way it's done atm but I can't think of anything better. I think that's just the nature of concurrency..